### PR TITLE
Event brite

### PR DESCRIPTION
**Pull request by erezfree29**
Project: Private Events

The App allows users to create events and then manage user signups. 

The app uses a 'sessions controller' and sessions to manage user sign up/in/out and avoids using the devise gem.

Users can create events 

Events take place at a specific date and at a location.

A user can create events. A user can attend many events. 

An event can be attended by many users. 

A user can see all events and make a booking for future events.

A user can see events he attended in the past and future events he is listed on.

The app will prevent a user from rebooking to an event he is already listed on.

### CODE REVIEW

File: app/views/events/_all_events.html.erb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  <h1>future events</h1>
+  <%@events.each_with_index do |event,index|%>
+    <% if !event.date.past?%>
+      <h1><%=index + 1%>) Name: <%=event.name%> </h1>
+      <h2>Location: <%=event.location%></h2>
+      <h2>Date: <%=event.date%> </h1>
+      <%=link_to 'book',"/attend/#{event.id}",method: :post%>
+    <% end %>  
+  <% end %>
+<% end %>
```
Comment (oloomoses) [inline app/views/events/_all_events.html.erb]: - [ ] Kindly move this logic into helper

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -11,6 +11,16 @@
...
+  <div class="nav_bar">
+     <%=link_to 'Home', root_path%>
+    <%if logged_in?%>
+        <h1><%=current_user.name%>-logged in</h1>
+      <% end %>
+      <%= button_to "Sign up",new_user_path, method: :get if (!logged_in? && controller_name != 'users')%><br>
+      <%= button_to "Log in",new_session_path, method: :get if !logged_in?%><br>
+      <%= button_to "Sign out",sign_out_path, method: :delete if logged_in?%><br>
+  </div>
+    <%= yield %><br>
```
Comment (oloomoses) [inline app/views/layouts/application.html.erb]: - [ ] Kindly move this logic into helpers

File: app/views/users/_user_events.html.erb [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  
+  <h1>Future Events</h1>
+  <%@events_ids_array.each_with_index do |event_id,index|%>
+  <% if  Event.all[event_id - 1].date > Time.now%>
+    <h1><%= index + 1 %>) name: <%=Event.all[event_id - 1].name %></h1>
+     <h2>Location: <%= Event.all[event_id - 1].location %></h2>
+    <h2>Date: <%= Event.all[event_id - 1].date %></h2>
+    <% end %>
+  <% end %>
+<% end %>
```
Comment (oloomoses) [inline app/views/users/_user_events.html.erb]: - [ ] Kindly move this logic into helpers

**Review by oloomoses — CHANGES_REQUESTED** [review]
# Changes Required :recycle:
**Hello @erezfree29, great job on this project so far.**

## Highlights
- Well done on the rubocop and linters configuration :muscle: 
- Great job on the pull request and the commit messages :100: 
- Impressive work on the login page and the authentication :100: 

# To Improve
- [ ] Kindly add validation to the event, to prevent a user from creating an event with empty fields
- [ ] When I am not Logged in, I can see a `create event` link. Kindly ensure that only a Logged in user can create a new event ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_can_create_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [ ] Kindly fix the error generated when the events page is visited. Event's index page should display past and upcoming events lists ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [ ] The event's show page displays the list of attendants ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/events_attendees_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [ ] The user's show page displays events the user has created and events that the user is attending to separated in past and upcoming ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [ ] Kindly add the Navigation links at the top of the page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/navigation_links_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

#### Models
- [ ] Kindly add scopes for the upcoming and the past events in the events model. Check the example below :arrow_down: 
```
scope :upcoming, -> { where('date >= ?', Time.zone.now) }
scope :previous, -> { where('date < ?', Time.zone.now) }
```
- [ ] Kindly add a `foreign key` for the events association in the user model. example: `has_many :events, foreign_key: :creator_id`
- [ ] Kindly create an association between event_etendee(user) and the event to enable a user to attend an event. [Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/attendee_assoc_requirement.png)

#### Controllers
##### EventsController
###### `create` action
- [ ] Use #build method on association, i.e. `current_user.events.build(event_params)` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/build_method_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

###### `index` action
* [ ] lists past events, i.e. `@past_events = Event.previous_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] lists ucomping events, i.e. `@upcoming_events = Event.upcoming_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

##### UsersController
###### `show` action

* [ ] lists events created by user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

#### Views
- [ ] There are logics in your views that should be moved into helpers. views should strictly be used to render content from the corresponding controller. Kindly check the comments. Check out this article to get started with [rails helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)

#### Test
* [ ] Kindly ensure that the models have unit tests for validations and associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
* [ ] Kindly ensure that the authentication have has integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
* [ ] Kindly ensure that the event managment features have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))


Kindly implement the above changes before submitting another review.
**Feel free to reach me on slack or leave a comment while mentioning me and I will be happy to reply**

_Slack: Oloo Moses_

**Comment by erezfree29** [comment]
I have made the suggested changes except.
1)I have moved some of the logic from the views except for each loops the reason is that 
I have found that the helpers do not handle arrays well and instead of looping over it they return the entire array.

I have not done testing as it is not required in this project ,I have consulted with my stand up meeting on this matter and have been advised that their project was approved without testing and I would like to move forward , I may however may add testing at a later stage.

**Comment by erezfree29** [comment]
I have done the required changes.
1)I have moved the logic from the application.html.erb to the application helper.
2)I have not moved the logic from the views which included arrays that involved iterations for two reasons.
firstly I have encountered an issue with iterating over arrays in the helpers as they returned the arrays instead of elements inside them, secondly, all the logic in my views is in rendered views and that does not make the code more difficult to read or any less clean.
3)I have not done testing as this was not required in the project, I have consulted on the matter with my standup team and they have advised me that their projects were approved without the testing part. I may however add testing at a later stage but at the moment I would prefer to save time and move on to the next project.

**Review by oloomoses — CHANGES_REQUESTED** [review]
# Changes Required :recycle: 
**Hello @erezfree29, great job working on some of the suggested changes, you have made great improvements since the last review. Some changes to be made before the project can be approved.**

- [ ] Testing is a requirement for this project. Kindly check the [specific requirements](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/) for more details.
- [ ] It is important that you move all your logic into helpers if you must have them. Otherwise, most of the conditional logic in your views are unnecessary.
### Events Controller
- [ ] In the index action you don't have to query all database, instead, you can just pull the past and the previous events like:
```
    @upcoming_events = Event.upcoming
    @past_events = Event.past
```
This way you don't have to create conditional logic in the view since you can access the events directly from the controller.

### users controller
##### `show action`
Since you have an association between event_attendee(user) and that event, It is recommended that you list the attended_events as suggested bellow
* [ ] lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

Kindly implement the above changes before submitting another review.
**Feel free to reach me on slack or leave a comment while mentioning me and I will be happy to reply**

_Slack: Oloo Moses_

**Review by oloomoses — APPROVED** [review]
# Congratulations, Project Approved. :muscle: :tada: :clap: 
**Hello @erezfree29 
Great job working on the suggested changes from the previous review.
You have passed all the requirements for this project :muscle: 
Feel free to merge and close this pull request.**

**I wish you the best in your next project and see you at the top :rocket:** 
#### Happy Coding
_Slack: Oloo Moses_

**Pull request merged**